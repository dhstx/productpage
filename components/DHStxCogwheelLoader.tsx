// components/DHStxCogwheelLoader.tsx
"use client";
import { useEffect, useRef } from "react";
import { animate } from "animejs";

type Speed = "slow" | "normal" | "fast";
type Size = "sm" | "md" | "lg";

const SIZE_MAP: Record<Size, number> = { sm: 80, md: 120, lg: 160 };
const DUR_MAP: Record<Speed, { big: number; small: number }> = {
  slow:   { big: 4000, small: 3000 },
  normal: { big: 3000, small: 2000 },
  fast:   { big: 2000, small: 1000 },
};

export function DHStxCogwheelLoader({
  text = "Processing...",
  size = "md",
  speed = "normal",
}: { text?: string; size?: Size; speed?: Speed }) {
  const bigRef = useRef<SVGSVGElement>(null);
  const smallRef = useRef<SVGSVGElement>(null);
  const container = SIZE_MAP[size];
  const { big, small } = DUR_MAP[speed];

  useEffect(() => {
    const media = window.matchMedia("(prefers-reduced-motion: reduce)");
    if (media.matches) return;

    const a1 = animate(bigRef.current, {
      rotate: 360,
      duration: big,
      ease: "linear",
      loop: true,
      autoplay: true,
    });
    const a2 = animate(smallRef.current, {
      rotate: -360,
      duration: small,
      ease: "linear",
      loop: true,
      autoplay: true,
    });
    return () => { a1.pause(); a2.pause(); };
  }, [big, small]);

  return (
    <div className="flex flex-col items-center justify-center space-y-4" role="status" aria-live="polite">
      <div className="relative" style={{ width: container, height: container }}>
        {/* Large Gear */}
        <svg ref={bigRef} className="absolute inset-0 m-auto w-4/5 h-4/5" viewBox="0 0 100 100" aria-hidden="true">
          <circle cx="50" cy="50" r="20" fill="#1A1A1A" stroke="#FFC96C" strokeWidth="2" />
          <circle cx="50" cy="50" r="35" fill="none" stroke="#FFC96C" strokeWidth="2" />
          {Array.from({ length: 8 }).map((_, i) => {
            const angle = (i * 360) / 8; const rad = (angle * Math.PI) / 180;
            const x1 = 50 + 35 * Math.cos(rad); const y1 = 50 + 35 * Math.sin(rad);
            const x2 = 50 + 45 * Math.cos(rad); const y2 = 50 + 45 * Math.sin(rad);
            return <line key={i} x1={x1} y1={y1} x2={x2} y2={y2} stroke="#FFC96C" strokeWidth="6" />;
          })}
        </svg>
        {/* Small Gear */}
        <svg ref={smallRef} className="absolute top-0 right-0 w-1/3 h-1/3" viewBox="0 0 100 100" aria-hidden="true">
          <circle cx="50" cy="50" r="15" fill="#1A1A1A" stroke="#FFC96C" strokeWidth="2" />
          <circle cx="50" cy="50" r="30" fill="none" stroke="#FFC96C" strokeWidth="2" />
          {Array.from({ length: 6 }).map((_, i) => {
            const angle = (i * 360) / 6; const rad = (angle * Math.PI) / 180;
            const x1 = 50 + 30 * Math.cos(rad); const y1 = 50 + 30 * Math.sin(rad);
            const x2 = 50 + 40 * Math.cos(rad); const y2 = 50 + 40 * Math.sin(rad);
            return <line key={i} x1={x1} y1={y1} x2={x2} y2={y2} stroke="#FFC96C" strokeWidth="5" />;
          })}
        </svg>
      </div>
      {text !== "" && (
        <p className="font-mono text-xs text-[#B3B3B3]">
          <span className="text-[#FFC96C]">â–®</span> {text}
        </p>
      )}
    </div>
  );
}
