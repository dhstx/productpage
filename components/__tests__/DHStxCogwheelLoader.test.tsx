import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import { DHStxCogwheelLoader } from '../DHStxCogwheelLoader';

describe('DHStxCogwheelLoader', () => {
  beforeEach(() => {
    // Mock matchMedia for normal motion
    window.matchMedia = vi.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    }));
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('renders with default text', () => {
    const { container } = render(<DHStxCogwheelLoader />);
    expect(screen.getByText('Processing...')).toBeInTheDocument();
    expect(screen.getByRole('status')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders without text when text prop is empty string', () => {
    const { container } = render(<DHStxCogwheelLoader text="" />);
    expect(screen.queryByText('Processing...')).not.toBeInTheDocument();
    expect(screen.getByRole('status')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('respects prefers-reduced-motion', () => {
    // Mock matchMedia to return reduced motion preference
    window.matchMedia = vi.fn().mockImplementation(query => ({
      matches: query === '(prefers-reduced-motion: reduce)',
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn(),
    }));

    const { container } = render(<DHStxCogwheelLoader />);
    
    // Check that the component renders without errors
    expect(screen.getByRole('status')).toBeInTheDocument();
    
    // The component should not have inline transform styles when reduced motion is preferred
    // (anime.js won't be initialized, so no transforms will be applied)
    const svgElements = container.querySelectorAll('svg');
    expect(svgElements.length).toBe(2);
  });

  it('renders with custom size', () => {
    const { container } = render(<DHStxCogwheelLoader size="lg" />);
    const statusDiv = container.querySelector('[role="status"]');
    expect(statusDiv).toBeInTheDocument();
  });

  it('renders with custom speed', () => {
    const { container } = render(<DHStxCogwheelLoader speed="fast" />);
    const statusDiv = container.querySelector('[role="status"]');
    expect(statusDiv).toBeInTheDocument();
  });

  it('has proper accessibility attributes', () => {
    render(<DHStxCogwheelLoader />);
    const status = screen.getByRole('status');
    expect(status).toHaveAttribute('aria-live', 'polite');
  });
});
