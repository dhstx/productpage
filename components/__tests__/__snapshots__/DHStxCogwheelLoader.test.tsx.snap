// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DHStxCogwheelLoader > renders with default text 1`] = `
<div>
  <div
    aria-live="polite"
    class="flex flex-col items-center justify-center space-y-4"
    role="status"
  >
    <div
      class="relative"
      style="width: 120px; height: 120px;"
    >
      <svg
        aria-hidden="true"
        class="absolute inset-0 m-auto w-4/5 h-4/5"
        style="transform: rotate(0deg);"
        viewBox="0 0 100 100"
      >
        <circle
          cx="50"
          cy="50"
          fill="#1A1A1A"
          r="20"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="35"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="85"
          x2="95"
          y1="50"
          y2="50"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="74.74873734152916"
          x2="81.81980515339464"
          y1="74.74873734152916"
          y2="81.81980515339464"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="50"
          x2="50"
          y1="85"
          y2="95"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="25.25126265847084"
          x2="18.180194846605364"
          y1="74.74873734152916"
          y2="81.81980515339464"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="15"
          x2="5"
          y1="50.00000000000001"
          y2="50.00000000000001"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="25.251262658470832"
          x2="18.180194846605353"
          y1="25.25126265847084"
          y2="18.180194846605364"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="49.99999999999999"
          x2="49.99999999999999"
          y1="15"
          y2="5"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="74.74873734152916"
          x2="81.81980515339463"
          y1="25.251262658470832"
          y2="18.180194846605353"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="absolute top-0 right-0 w-1/3 h-1/3"
        style="transform: rotate(0deg);"
        viewBox="0 0 100 100"
      >
        <circle
          cx="50"
          cy="50"
          fill="#1A1A1A"
          r="15"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="30"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="80"
          x2="90"
          y1="50"
          y2="50"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="65"
          x2="70"
          y1="75.98076211353316"
          y2="84.64101615137754"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="35.00000000000001"
          x2="30.000000000000007"
          y1="75.98076211353316"
          y2="84.64101615137756"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="20"
          x2="10"
          y1="50.00000000000001"
          y2="50.00000000000001"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="34.999999999999986"
          x2="29.999999999999982"
          y1="24.019237886466847"
          y2="15.358983848622458"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="65"
          x2="70"
          y1="24.019237886466843"
          y2="15.358983848622458"
        />
      </svg>
    </div>
    <p
      class="font-mono text-xs text-[#B3B3B3]"
    >
      <span
        class="text-[#FFC96C]"
      >
        â–®
      </span>
       
      Processing...
    </p>
  </div>
</div>
`;

exports[`DHStxCogwheelLoader > renders without text when text prop is empty string 1`] = `
<div>
  <div
    aria-live="polite"
    class="flex flex-col items-center justify-center space-y-4"
    role="status"
  >
    <div
      class="relative"
      style="width: 120px; height: 120px;"
    >
      <svg
        aria-hidden="true"
        class="absolute inset-0 m-auto w-4/5 h-4/5"
        style="transform: rotate(0deg);"
        viewBox="0 0 100 100"
      >
        <circle
          cx="50"
          cy="50"
          fill="#1A1A1A"
          r="20"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="35"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="85"
          x2="95"
          y1="50"
          y2="50"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="74.74873734152916"
          x2="81.81980515339464"
          y1="74.74873734152916"
          y2="81.81980515339464"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="50"
          x2="50"
          y1="85"
          y2="95"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="25.25126265847084"
          x2="18.180194846605364"
          y1="74.74873734152916"
          y2="81.81980515339464"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="15"
          x2="5"
          y1="50.00000000000001"
          y2="50.00000000000001"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="25.251262658470832"
          x2="18.180194846605353"
          y1="25.25126265847084"
          y2="18.180194846605364"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="49.99999999999999"
          x2="49.99999999999999"
          y1="15"
          y2="5"
        />
        <line
          stroke="#FFC96C"
          stroke-width="6"
          x1="74.74873734152916"
          x2="81.81980515339463"
          y1="25.251262658470832"
          y2="18.180194846605353"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="absolute top-0 right-0 w-1/3 h-1/3"
        style="transform: rotate(0deg);"
        viewBox="0 0 100 100"
      >
        <circle
          cx="50"
          cy="50"
          fill="#1A1A1A"
          r="15"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="30"
          stroke="#FFC96C"
          stroke-width="2"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="80"
          x2="90"
          y1="50"
          y2="50"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="65"
          x2="70"
          y1="75.98076211353316"
          y2="84.64101615137754"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="35.00000000000001"
          x2="30.000000000000007"
          y1="75.98076211353316"
          y2="84.64101615137756"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="20"
          x2="10"
          y1="50.00000000000001"
          y2="50.00000000000001"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="34.999999999999986"
          x2="29.999999999999982"
          y1="24.019237886466847"
          y2="15.358983848622458"
        />
        <line
          stroke="#FFC96C"
          stroke-width="5"
          x1="65"
          x2="70"
          y1="24.019237886466843"
          y2="15.358983848622458"
        />
      </svg>
    </div>
  </div>
</div>
`;
